diff --git a/main/socat/APKBUILD b/main/socat/APKBUILD
index ba756448..10342f08 100644
--- a/main/socat/APKBUILD
+++ b/main/socat/APKBUILD
@@ -19,11 +19,11 @@ source="http://www.dest-unreach.org/socat/download/socat-$pkgve
r.tar.gz
        "
 
 build() {
-       ./configure \
+       ./configure LDFLAGS=-static \
                --build=$CBUILD \
                --host=$CHOST \
                --prefix=/usr
-       make
+       make LDFLAGS=-static
 }
 
 check() {

