diff --git a/main/strace/APKBUILD b/main/strace/APKBUILD
index bee37082..86bd0b37 100644
--- a/main/strace/APKBUILD
+++ b/main/strace/APKBUILD
@@ -17,7 +17,11 @@ source="https://github.com/strace/strace/releases/download/v$pkgver/strace-$pkgv
 	"
 
 build() {
-	CFLAGS="$CFLAGS -flto=auto" \
+	CFLAGS="$CFLAGS -flto=auto -g -O0" \
+	CPPFLAGS=-DDEBUG \
+	LDFLAGS=-static \
+	LIBS="-l:liblzma.a -l:libucontext.a" \
+	PKG_CONFIG="pkg-config --static" \
 	./configure \
 		--build=$CBUILD \
 		--host=$CHOST \
@@ -25,6 +29,7 @@ build() {
 		--sysconfdir=/etc \
 		--mandir=/usr/share/man \
 		--localstatedir=/var \
+		--with-libunwind \
 		--enable-mpers=no
 	make
 }
